File: frontend/src/types/attendance.ts
export interface Attendance {
  id: number;
  employee_id: number;
  date: string;
  entry_time: string;
  exit_time: string | null;
  hours_worked: number | string | null;
  created_at: string;
  updated_at: string;
}

export interface DailyEntry {
  date: string;
  present_count: number;
  absent_count: number;
  late_count: number;
}

export interface DashboardOverview {
  month: number;
  year: number;
  total_employees: number;
  working_days: number;
  daily_entries: DailyEntry[];
}

export interface EmployeeStat {
  employee_id: number;
  name: string;
  shift: string;
  days_present: number;
  days_absent: number;
  attendance_rate: number;
  late_days: number;
  overtime_hours: number;
}

export interface DailyEmployeeStatus {
  employee_id: number;
  name: string;
  shift: string;
  status: string;
  entry_time: string | null;
  exit_time: string | null;
  hours_worked: number | null;
  is_late: boolean;
  overtime_hours: number;
}

export interface MonthlyReport {
  employee_id: number;
  employee_name: string;
  month: number;
  year: number;
  total_days: number;
  total_hours: number | string;
  records: Attendance[];
}

export interface FaceScanResult {
  employee_id: number;
  employee_name: string;
  action: string;
  attendance: Attendance;
}

export type MissedReason = "absent" | "incomplete" | "late";

export interface MissedDayDetail {
  date: string;
  reason: MissedReason;
  entry_time: string | null;
  exit_time: string | null;
}

export interface MissedAttendanceEntry {
  employee_id: number;
  employee_code: string | null;
  name: string | null;
  shift: string | null;
  missed_days: number;
  absent_days: number;
  incomplete_days: number;
  late_days: number;
  details: MissedDayDetail[];
}

export interface MissedAttendanceResponse {
  period: "daily" | "weekly" | "monthly";
  start_date: string;
  end_date: string;
  working_days: number;
  total_employees: number;
  employees_with_misses: number;
  total_absent: number;
  total_incomplete: number;
  total_late: number;
  total_missed: number;
  employees: MissedAttendanceEntry[];
}

File: backend/schemas/attendance.py
from pydantic import BaseModel
from decimal import Decimal
from datetime import date, time, datetime
from typing import Optional


class AttendanceClockIn(BaseModel):
    employee_id: int
    date:        date
    entry_time:  time
    image:       str        # base64 face photo — required for verification
    latitude:    Optional[float] = None
    longitude:   Optional[float] = None


class AttendanceClockOut(BaseModel):
    exit_time:  time
    image:      str        # base64 face photo — required for verification
    latitude:   Optional[float] = None
    longitude:  Optional[float] = None


class AttendanceManualClockIn(BaseModel):
    """Clock-in without face verification (fallback when face capture is unavailable)."""
    employee_id: int
    date:        date
    entry_time:  time
    latitude:    Optional[float] = None
    longitude:   Optional[float] = None


class AttendanceManualClockOut(BaseModel):
    """Clock-out without face verification."""
    exit_time:  time
    latitude:   Optional[float] = None
    longitude:  Optional[float] = None


class AttendanceUpdate(BaseModel):
    date:       Optional[date] = None
    entry_time: Optional[time] = None
    exit_time:  Optional[time] = None


class AttendanceResponse(BaseModel):
    id:                  int
    employee_id:         int
    date:                date
    entry_time:          time
    exit_time:           Optional[time]
    hours_worked:        Optional[Decimal]
    clock_in_latitude:   Optional[float] = None
    clock_in_longitude:  Optional[float] = None
    clock_out_latitude:  Optional[float] = None
    clock_out_longitude: Optional[float] = None
    created_at:          datetime
    updated_at:          datetime

    model_config = {"from_attributes": True}


class MonthlyAttendanceSummary(BaseModel):
    employee_id:   int
    employee_name: Optional[str] = None
    month:         int
    year:          int
    total_days:    int
    total_hours:   Decimal
    records:       list[AttendanceResponse]


# ── Dashboard schemas ────────────────────────────────────────────────────────

class DailyOverviewEntry(BaseModel):
    date:            date
    present_count:   int
    absent_count:    int
    late_count:      int
    total_employees: int


class DashboardOverviewResponse(BaseModel):
    month:           int
    year:            int
    total_employees: int
    working_days:    int
    daily_entries:   list[DailyOverviewEntry]


class EmployeeStatEntry(BaseModel):
    employee_id:     int
    name:            Optional[str] = None
    shift:           Optional[str] = None
    days_present:    int
    days_absent:     int
    attendance_rate: float
    late_days:       int
    overtime_hours:  float


class DailyEmployeeStatus(BaseModel):
    employee_id:    int
    name:           str
    shift:          str
    status:         str   # "present" | "absent" | "holiday"
    entry_time:     Optional[time]
    exit_time:      Optional[time]
    hours_worked:   Optional[Decimal]
    is_late:        bool
    overtime_hours: float



# ── Missed attendance schemas ────────────────────────────────────────────────

class MissedDayDetail(BaseModel):
    date:       date
    reason:     str   # "absent" | "incomplete" | "late"
    entry_time: Optional[time] = None
    exit_time:  Optional[time] = None


class MissedAttendanceEntry(BaseModel):
    employee_id:     int
    employee_code:   Optional[str] = None
    name:            Optional[str] = None
    shift:           Optional[str] = None
    missed_days:     int
    absent_days:     int
    incomplete_days: int
    late_days:       int
    details:         list[MissedDayDetail]


class MissedAttendanceResponse(BaseModel):
    period:                str   # "daily" | "weekly" | "monthly"
    start_date:            date
    end_date:              date
    working_days:          int
    total_employees:       int
    employees_with_misses: int
    total_absent:          int
    total_incomplete:      int
    total_late:            int
    total_missed:          int
    employees:             list[MissedAttendanceEntry]
